spring-cloud-stream/src/main/java/org/springframework/cloud/stream/config/SmartMessageMethodArgumentResolver.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2016-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.stream.config;
+
+import java.lang.reflect.Type;
+
+import org.springframework.core.MethodParameter;
+import org.springframework.core.ResolvableType;
+import org.springframework.lang.Nullable;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.converter.MessageConversionException;
+import org.springframework.messaging.converter.MessageConverter;
+import org.springframework.messaging.converter.SmartMessageConverter;
+import org.springframework.messaging.handler.annotation.support.MessageMethodArgumentResolver;
+import org.springframework.messaging.handler.annotation.support.MethodArgumentTypeMismatchException;
+import org.springframework.messaging.support.MessageBuilder;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.StringUtils;
+
+/**
+ * 
+ * @author Oleg Zhurakousky
+ * 
+ * @deprecated will be removed once https://jira.spring.io/browse/SPR-17503 is addressed
+ */
+@Deprecated
+class SmartMessageMethodArgumentResolver extends MessageMethodArgumentResolver {
+	
+	private final MessageConverter messageConverter;
+	
+	SmartMessageMethodArgumentResolver() {
+		this(null);
+	}
+
+	/**
+	 * Create a resolver instance with the given {@link MessageConverter}.
+	 * @param converter the MessageConverter to use (may be {@code null})
+	 * @since 4.3
+	 */
+	SmartMessageMethodArgumentResolver(@Nullable MessageConverter converter) {
+		this.messageConverter = converter;
+	}
+
+	@Override
+	public Object resolveArgument(MethodParameter parameter, Message<?> message) throws Exception {
+		Class<?> targetMessageType = parameter.getParameterType();
+		Class<?> targetPayloadType = getPayloadType(parameter);
+
+		if (!targetMessageType.isAssignableFrom(message.getClass())) {
+			throw new MethodArgumentTypeMismatchException(message, parameter, "Actual message type '" +
+					ClassUtils.getDescriptiveType(message) + "' does not match expected type '" +
+					ClassUtils.getQualifiedName(targetMessageType) + "'");
+		}
+
+		
+		Class<?> targetClass = parameter.getParameterType();
+		Class<?> payloadClass = message.getPayload().getClass();
+
+		if (ClassUtils.isAssignable(payloadClass, targetClass)) {
```

Comment:
?? `targetClass == targetMessageType`; unlikely that the payload is a message.

Shouldn't this check be testing whether the parameter payload type (e.g. `Foo` in `Message<Foo>` ) is assignable from the actual payload type? (which is what the standard resolver is doing).

I can't get my head around `isAssignable(actualClass, wantedClass)`; surely it's `isAssignable(wantedClass, actualClass)`.